### PR TITLE
Add path @at for positional indexing, and negative indices

### PR DESCRIPTION
**Pull request by kalman**
This lets you do foo.bar@at(n) to get the nth element of a list, set, or
map (for lists, this is equivalent to foo.bar[n]). This patch also adds
support for negative indices to @at(-n) and [-n] to get the nth element
relative to the back of collections.

### CODE REVIEW

File: go/types/path.go [hunk 1]
```diff
@@ -17,7 +17,10 @@ import (
 	"github.com/attic-labs/noms/go/hash"
 )
 
-var annotationRe = regexp.MustCompile("^([a-z]+)")
+// For an annotation like @type, 1st capture group is the annotation.
+// For @at(42), 1st capture group is the annotation and 3rd is the parameter.
+// Note, @at() is valid under this regexp, code should deal with the error.
```
Comment (aboodman) [inline go/types/path.go]: OOC, why not make `at()` invalid?

Comment (kalman) [inline go/types/path.go]: So I can distinguish between `@at` and `@at()` for constructing error messages, i.e. "missing annotation" vs "not a number".

**Comment by aboodman** [comment]
Makes sense. LGTM, modulo doc.

**Comment by aboodman** [comment]
LGTM

**Pull request merged**